components/server/ee/src/workspace/gitpod-server-impl.ts

```diff
@@ -2047,25 +2048,44 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
         }
     }
 
-    async findStripeSubscriptionIdForTeam(ctx: TraceContext, teamId: string): Promise<string | undefined> {
-        this.checkAndBlockUser("findStripeSubscriptionIdForTeam");
-        await this.guardTeamOperation(teamId, "get");
+    async findStripeSubscriptionId(ctx: TraceContext, attributionId: string): Promise<string | undefined> {
+        const attrId = AttributionId.parse(attributionId);
+        if (attrId === undefined) {
+            log.error(`Invalid attribution id: ${attributionId}`);
+            throw new ResponseError(ErrorCodes.BAD_REQUEST, `Invalid attibution id: ${attributionId}`);
+        }
+
+        this.checkAndBlockUser("findStripeSubscriptionId");
+
+        let customer: Stripe.Customer | undefined;
         try {
-            const customer = await this.stripeService.findCustomerByTeamId(teamId);
+            if (attrId.kind == "team") {
+                await this.guardTeamOperation(attrId.teamId, "get");
+                customer = await this.stripeService.findCustomerByTeamId(attrId.teamId);
+            } else {
```

Comment:
Perhaps worth being explcit about `user` in case we add more types of attribution ids.

Reply:
Would it? Here we just do a check against `== "team"` and then do an `else` for anything else. I think it wouldn't fail to compile in that case.

Reply:
If I extend the `AttributionId` union with a new `"foo"` type as below:

<summary>
<details>

```ts
export type AttributionId = UserAttributionId | TeamAttributionId | FooAttributionId;
export type AttributionTarget = "user" | "team" | "foo";

export interface UserAttributionId {
    kind: "user";
    userId: string;
}
export interface TeamAttributionId {
    kind: "team";
    teamId: string;
}

export interface FooAttributionId {
    kind: "foo";
    someOtherId: string;
}

export namespace AttributionId {
    const SEPARATOR = ":";

    export function parse(s: string): UserAttributionId | TeamAttributionId | FooAttributionId | undefined {
        if (!s) {
            return undefined;
        }
        const parts = s.split(":");
        if (parts.length !== 2) {
            return undefined;
        }
        switch (parts[0]) {
            case "user":
                return { kind: "user", userId: parts[1] };
            case "team":
                return { kind: "team", teamId: parts[1] };
            case "foo":
                return { kind: "foo", someOtherId: parts[1] };
            default:
                return undefined;
        }
    }

    export function render(id: AttributionId): string {
        switch (id.kind) {
            case "user":
                return `user${SEPARATOR}${id.userId}`;
            case "team":
                return `team${SEPARATOR}${id.teamId}`;
            case "foo":
                return `foo{SEPARATOR}${id.someOtherId}`;
        }
    }
}
```
</details>
</summary>

I get a type error in the `else` block here:

https://github.com/gitpod-io/gitpod/blob/015987aa4288fe23ff2fc846209da6671b06764d/components/server/ee/src/workspace/gitpod-server-impl.ts#L2066

because the type of `attrId` in the `else` block is now `UserAttributionId | FooAttributionId` rather than `UserAttributionId` and so can't be guaranteed to have a `userId` field.

Reply:
This is quite a neat feature of typescript's discriminated unions:

https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions

Reply:
When I do this in a [playground](https://www.typescriptlang.org/play?#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCbvCwDKgU8JkY7p7ehCTkVDQS2E6gnPCxGcwmZqDSTgzxxWWVoASMFmgYkAAeRJTInN3ymj4d-jSCeNsMq-wuoPaOltigAKoASgAywhK7SbGQZIIz5VWCFzSeCrZagNYbChbHaxUDcCjJZLfSDbExIAgUdxkUBIursJzCFJtXydajBRSodqVVCoSjSdj1RoASVcAF4HoJvhUaXSGQ1OCzQAAfbweRDc2kUemMgWuEUAb1A8Qy3HyoAARMl4PB1aAAL4AbmCGQUZGSDCW905ZAlvJlgvlihKJWVPDV6vY1vVRudoE93xZasEks4zCNesUJu+5qWODFtqlfOZrkdvtdqo1CgYiG9TudWcQgauIbDigjimSXFgMpx2oAFHVaWqE9L+SyAJSgVMiVINnkBdOgVnDzNi9Wd7slBCcQTwdwBaDwZh19UF8d5vXg6Ccrt5qeNWfzxfLj0JAYSTjrkoR8sV+uK9PurU6-XtxRAA), it compiles

Reply:
But if you try to do anything with the value of `attr` inside your `else` block [like access `userId`](https://www.typescriptlang.org/play?#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCbvCwDKgU8JkY7p7ehCTkVDQS2E6gnPCxGcwmZqDSTgzxxWWVoASMFmgYkAAeRJTInN3ymj4d-jSCeNsMq-wuoPaOltigAKoASgAywhK7SbGQZIIz5VWCFzSeCrZagNYbChbHaxUDcCjJZLfSDbExIAgUdxkUBIursJzCFJtXydajBRSodqVVCoSjSdj1RoASVcAF4HoJvhUaXSGQ1OCzQAAfbweRDc2kUemMgWuEUAb1A8Qy3HyoAARMl4PB1aAAL4AbmCGQUZGSDCW905ZAlvJlgvlihKJWVPDV6vY1vVRudoE93xZasEks4zCNesUJu+5qWODFtqlfOZrkdvtdqo1CgYiG9TudWcQgauIbDigjimSXFgMpx2oAFHVaWqE9L+SyAJSgVMiVINnkBdOgVnDzNi9Wd7slBCcQTwdwBaDwZh92kBAsdvN68HQTldvNTxqz+eL5eNsgBf1kDclCPliv1xXp91anX69uKIA), you do get a type error because the type has not been sufficiently narrowed inside the `else` (it could still be either a `foo` or a `user`).

Reply:
Sure, but that relies on the fact that we wouldn't overload the `userId` or another property. If my new type was for a `{ kind: "superuser", userId: "..."}` then it will compile, but won't work as expected.

Reply:
yeah, that's true. I assume that if we added a new attribution `kind` it would have a non-overlapping shape.